File: LeetCode/Easy/0383-ransom-note/0383-ransom-note.go
func canConstruct(ransomNote string, magazine string) bool {
    m := make(map[rune]int)

    for _, char := range magazine {
        _, exists := m[char]
        if (!exists) {
            m[char] = 1
        } else {
            m[char] += 1
        }
    }

    for _, char := range ransomNote {
        val, exists := m[char]
        if (!exists || val == 0) {
            return false
        } else {
            m[char] -= 1
        }
    }


    return true
}

File: LeetCode/Easy/0383-ransom-note/0383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> rMap = new HashMap<>();
        HashMap<Character, Integer> mMap = new HashMap<>();

        for (int i = 0; i < ransomNote.length(); i++) {
            Character key = ransomNote.charAt(i);
            rMap.put(key, rMap.getOrDefault(key, 0) + 1);
        }

        for (int i = 0; i < magazine.length(); i++) {
            Character key = magazine.charAt(i);
            mMap.put(key, mMap.getOrDefault(key, 0) + 1);
        }

        for (Character key: rMap.keySet()) {
            if (rMap.get(key) > mMap.getOrDefault(key, 0)) {
                return false;
            }
        }
        return true;
    }
}